api/api-environments-v3/src/main/java/com/thoughtworks/go/apiv3/environments/EnvironmentsControllerV3.java

```diff
@@ -0,0 +1,268 @@
+            return handleCreateOrUpdateResponse(request, response, environmentConfigToCreate, operationResult);
+        }
+
+        haltIfEntityWithSameNameExists(environmentConfigToCreate);
+
+        environmentConfigService.createEnvironment(environmentConfigToCreate, currentUsername(), operationResult);
+
+        setEtagHeader(environmentConfigToCreate, response);
+        return handleCreateOrUpdateResponse(request, response, environmentConfigToCreate, operationResult);
+    }
+
+    public String update(Request request, Response response) {
+        String environmentName = request.params("name");
+        BasicEnvironmentConfig environmentConfig = (BasicEnvironmentConfig) buildEntityFromRequestBody(request);
+        EnvironmentConfig oldEnvironmentConfig = fetchEntityFromConfig(environmentName);
+        HttpLocalizedOperationResult operationResult = new HttpLocalizedOperationResult();
+
+        if (!environmentConfig.getAllErrors().isEmpty()) {
+            operationResult.unprocessableEntity("Error parsing environment config from the request");
+            return handleCreateOrUpdateResponse(request, response, environmentConfig, operationResult);
+        }
+
+        if (isPutRequestStale(request, oldEnvironmentConfig)) {
+            throw haltBecauseEtagDoesNotMatch("environment", environmentName);
+        }
+
+        if (!StringUtils.equals(environmentName, environmentConfig.name().toString())) {
+            throw haltBecauseRenameOfEntityIsNotSupported("environment");
+        }
+
+        environmentConfigService.updateEnvironment(environmentConfig.name().toString(), environmentConfig, currentUsername(), etagFor(oldEnvironmentConfig), operationResult);
+
+        setEtagHeader(environmentConfig, response);
+        return handleCreateOrUpdateResponse(request, response, environmentConfig, operationResult);
+    }
+
+    public String partialUpdate(Request request, Response response) {
+        JsonReader jsonReader = GsonTransformer.getInstance().jsonReaderFrom(request.body());
+        PatchEnvironmentRequest req = PatchEnvironmentRequestRepresenter.fromJSON(jsonReader);
+        HttpLocalizedOperationResult result = new HttpLocalizedOperationResult();
+
+        String environmentName = request.params("name");
+        EnvironmentConfig environmentConfig = fetchEntityFromConfig(environmentName);
+
+        Optional<EnvironmentVariableConfig> parsingErrors = req.getEnvironmentVariablesToAdd().stream()
+                .filter(envVar -> !envVar.errors().isEmpty())
+                .findFirst();
+
+        if (parsingErrors.isPresent()) {
+            EnvironmentVariablesConfig configs = new EnvironmentVariablesConfig(req.getEnvironmentVariablesToAdd());
+            String errorMessage = MessageJson.create("Error parsing patch request",
+                    writer -> toJSONArray(writer, "environment_variables", configs));
+            response.status(422);
+            return errorMessage;
```

Comment:
Will this send a string response?
How about using `MessageJson.create`?

Reply:
`errorMessage` is a reference to the result of `MessageJson.create`. A test is present to verify the string response.

Reply:
Ahh! Got it. May be inline the variable?